File: bao-source/bao-edge/Android/src/app/src/main/java/com/google/ai/edge/gallery/ui/home/OperatorAutomationExecutor.kt
package com.google.ai.edge.gallery.ui.home

import android.content.Context
import com.google.ai.edge.gallery.R
import com.google.ai.edge.gallery.common.StructuredLog
import com.google.ai.edge.gallery.data.AiWorkflowJobEnvelope
import com.google.ai.edge.gallery.data.AiWorkflowRequest
import com.google.ai.edge.gallery.data.CloudControlPlaneClient
import com.google.ai.edge.gallery.data.CredentialStore
import com.google.ai.edge.gallery.data.OperatorRuntimeUsage
import com.google.ai.edge.gallery.data.OperatorVerificationEvidence
import com.google.ai.edge.gallery.data.OperatorVerificationReplySummary
import com.google.ai.edge.gallery.data.OperatorVerificationReporter
import com.google.ai.edge.gallery.data.OperatorVerificationReports
import com.google.ai.edge.gallery.data.OperatorVerificationRequestSummary
import com.google.ai.edge.gallery.data.OperatorVerificationRuntimeBinding
import com.google.ai.edge.gallery.data.OperatorVerificationRuntimeTransport
import com.google.ai.edge.gallery.data.OperatorVerificationScenario
import com.google.ai.edge.gallery.data.OperatorVerificationTargetIds
import com.google.ai.edge.gallery.data.OperatorVerificationTerminalState
import com.google.ai.edge.gallery.data.OperatorVerificationTriggerSource
import com.google.ai.edge.gallery.data.OperatorVerificationUsage
import com.google.ai.edge.gallery.data.PreferencesStore
import dagger.Binds
import dagger.Module
import dagger.hilt.InstallIn
import dagger.hilt.android.qualifiers.ApplicationContext
import dagger.hilt.components.SingletonComponent
import java.time.Instant
import java.util.UUID
import javax.inject.Inject
import javax.inject.Singleton
import kotlinx.coroutines.delay
import kotlinx.coroutines.withContext
import kotlinx.coroutines.Dispatchers
import com.baohaus.baoedge.core.flow.FlowExecutionState

private const val TAG = "AGOperatorAutomationExecutor"
private const val AUTOMATION_MAX_POLLS = 120
private const val AUTOMATION_POLL_INTERVAL_MS = 2_000L

/** Terminal snapshot returned after a durable scheduled automation run. */
data class OperatorAutomationExecutionOutcome(
  val state: FlowExecutionState,
  val summary: String,
  val correlationId: String,
  val requiresAttention: Boolean,
)

/** Owner for executing scheduled automations outside the Android UI surface. */
interface OperatorAutomationExecutor {
  /** Execute one claimed scheduled automation and return the terminal outcome. */
  suspend fun execute(
    pendingRun: OperatorPendingScheduledRun,
  ): OperatorAutomationExecutionOutcome
}

@Singleton
internal class DefaultOperatorAutomationExecutor
@Inject
constructor(
  @ApplicationContext private val appContext: Context,
  private val cloudControlPlaneClient: CloudControlPlaneClient,
  private val credentialStore: CredentialStore,
  private val preferencesStore: PreferencesStore,
  private val nativeIntentExecutor: OperatorNativeIntentExecutor,
  private val verificationReporter: OperatorVerificationReporter,
) : OperatorAutomationExecutor {
  override suspend fun execute(
    pendingRun: OperatorPendingScheduledRun,
  ): OperatorAutomationExecutionOutcome =
    withContext(Dispatchers.IO) {
      val correlationId = UUID.randomUUID().toString()
      val startedAt = Instant.now().toString()

      val nativeResult = nativeIntentExecutor.executeIfSupported(prompt = pendingRun.prompt)
      if (nativeResult != null) {
        // Native intent executor writes its own verification report.
        return@withContext OperatorAutomationExecutionOutcome(
          state = nativeResult.state,
          summary = nativeResult.message,
          correlationId = correlationId,
          requiresAttention = nativeResult.state != FlowExecutionState.SUCCESS,
        )
      }

      val preferences = preferencesStore.readOperatorPreferences()
      val baseUrl = preferences.controlPlaneBaseUrl.trim()
      if (baseUrl.isEmpty()) {
        val errorSummary = appContext.getString(R.string.flow_runner_invalid_control_plane_base_url)
        persistVerificationReport(
          correlationId = correlationId,
          pendingRun = pendingRun,
          startedAt = startedAt,
          terminalState = OperatorVerificationTerminalState.ERROR_NON_RETRYABLE,
          summary = errorSummary,
          providerId = null,
          modelId = null,
        )
        return@withContext OperatorAutomationExecutionOutcome(
          state = FlowExecutionState.ERROR_NON_RETRYABLE,
          summary = errorSummary,
          correlationId = correlationId,
          requiresAttention = true,
        )
      }

      val assignment = preferences.runtimeAssignments.assignmentFor(OperatorRuntimeUsage.AUTOMATION)
      val providerId = assignment.provider.trim().ifEmpty { null }
      val modelId = assignment.model.trim().ifEmpty { null }
      val activeProfileId = providerId?.let { preferences.activeProfileIdFor(it).trim().ifEmpty { "default" } }
      val apiKey = if (providerId != null && activeProfileId != null) {
        credentialStore.readProviderApiKey(providerId = providerId, profileId = activeProfileId)?.trim()?.ifEmpty { null }
      } else {
        null
      }
      val providerBaseUrl = if (providerId != null) {
        preferences.activeProfileFor(providerId)?.baseUrl?.trim()?.ifEmpty { null }
      } else {
        preferences.providerBaseUrl.trim().ifEmpty { null }
      }

      StructuredLog.d(
        TAG,
        "operator_automation_schedule_execution_started",
        "scheduleId" to pendingRun.scheduleId,
        "provider" to providerId.orEmpty(),
        "model" to modelId.orEmpty(),
      )

      val envelope = runAgentWorkflow(
        baseUrl = baseUrl,
        prompt = pendingRun.prompt,
        providerId = providerId,
        modelId = modelId,
        apiKey = apiKey,
        providerBaseUrl = providerBaseUrl,
        correlationId = correlationId,
      )
      val summary = resolveWorkflowSummary(envelope)
      val state = resolveWorkflowState(envelope)
      persistVerificationReport(
        correlationId = correlationId,
        pendingRun = pendingRun,
        startedAt = startedAt,
        terminalState = OperatorVerificationTerminalState.fromFlowState(state),
        summary = summary,
        providerId = providerId,
        modelId = modelId,
      )
      OperatorAutomationExecutionOutcome(
        state = state,
        summary = summary,
        correlationId = correlationId,
        requiresAttention = state != FlowExecutionState.SUCCESS,
      )
    }

  private fun persistVerificationReport(
    correlationId: String,
    pendingRun: OperatorPendingScheduledRun,
    startedAt: String,
    terminalState: OperatorVerificationTerminalState,
    summary: String,
    providerId: String?,
    modelId: String?,
  ) {
    try {
      val report = OperatorVerificationReports.create(
        correlationId = correlationId,
        scenario = OperatorVerificationScenario.AUTOMATION_SCHEDULE,
        runtime = OperatorVerificationRuntimeBinding(
          usage = OperatorVerificationUsage.AUTOMATION.wireValue,
          transport = OperatorVerificationRuntimeTransport.CLOUD.wireValue,
          provider = providerId,
          model = modelId,
        ),
        triggerSource = OperatorVerificationTriggerSource.BACKGROUND_SCHEDULE,
        request = OperatorVerificationRequestSummary(prompt = pendingRun.prompt),
        terminalState = terminalState,
        reply = OperatorVerificationReplySummary(
          message = summary,
          state = terminalState.wireValue,
          provenance = "cloud",
        ),
        targetIds = OperatorVerificationTargetIds(scheduleId = pendingRun.scheduleId),
        startedAt = startedAt,
      )
      verificationReporter.persist(report)
    } catch (e: Exception) {
      StructuredLog.w(TAG, "operator_automation_verification_report_failed", "error" to e.message.orEmpty())
    }
  }

  private suspend fun runAgentWorkflow(
    baseUrl: String,
    prompt: String,
    providerId: String?,
    modelId: String?,
    apiKey: String?,
    providerBaseUrl: String?,
    correlationId: String,
  ): AiWorkflowJobEnvelope {
    var envelope =
      cloudControlPlaneClient.startAiWorkflowJob(
        baseUrl = baseUrl,
        request =
          AiWorkflowRequest(
            mode = "agent",
            provider = providerId,
            model = modelId,
            message = prompt,
            apiKey = apiKey,
            baseUrl = providerBaseUrl,
            correlationId = correlationId,
          ),
      )
    var pollCount = 0
    while (!isAiWorkflowJobTerminal(envelope) && pollCount < AUTOMATION_MAX_POLLS) {
      delay(AUTOMATION_POLL_INTERVAL_MS)
      envelope = cloudControlPlaneClient.getAiWorkflowJobEnvelope(baseUrl, envelope.jobId)
      pollCount += 1
    }
    return envelope
  }

  private fun isAiWorkflowJobTerminal(envelope: AiWorkflowJobEnvelope): Boolean {
    val status = envelope.data?.status ?: return true
    return status == "succeeded" || status == "failed" || status == "cancelled"
  }

  private fun resolveWorkflowState(envelope: AiWorkflowJobEnvelope): FlowExecutionState {
    return when (envelope.state) {
      "success" -> FlowExecutionState.SUCCESS
      "error-retryable" -> FlowExecutionState.ERROR_RETRYABLE
      "error-non-retryable" -> FlowExecutionState.ERROR_NON_RETRYABLE
      "unauthorized" -> FlowExecutionState.UNAUTHORIZED
      else ->
        if (envelope.data?.status == "succeeded") {
          FlowExecutionState.SUCCESS
        } else {
          FlowExecutionState.ERROR_NON_RETRYABLE
        }
    }
  }

  private fun resolveWorkflowSummary(envelope: AiWorkflowJobEnvelope): String {
    return envelope.data?.result?.reply?.trim().orEmpty().ifEmpty {
      envelope.data?.reason?.trim().orEmpty().ifEmpty {
        envelope.error?.reason.orEmpty().ifEmpty {
          envelope.mismatches.joinToString(separator = " ").ifEmpty {
            appContext.getString(R.string.operator_phone_execution_unavailable)
          }
        }
      }
    }
  }
}

@Module
@InstallIn(SingletonComponent::class)
internal abstract class OperatorAutomationExecutorModule {
  @Binds
  abstract fun bindOperatorAutomationExecutor(
    implementation: DefaultOperatorAutomationExecutor,
  ): OperatorAutomationExecutor
}

File: bao-source/bao-edge/iOS/BaoEdge/Sources/BaoEdgeDriver/OperatorAutomationExecutor.swift
import Foundation
import BaoEdgeCore

/// Terminal snapshot returned after a durable scheduled automation run.
public struct OperatorAutomationExecutionOutcome: Sendable {
    public let state: FlowExecutionState
    public let summary: String
    public let correlationId: String
    public let requiresAttention: Bool

    public init(
        state: FlowExecutionState,
        summary: String,
        correlationId: String,
        requiresAttention: Bool
    ) {
        self.state = state
        self.summary = summary
        self.correlationId = correlationId
        self.requiresAttention = requiresAttention
    }
}

/// Owner for executing scheduled automations outside the iOS UI surface.
public protocol OperatorAutomationExecuting: Sendable {
    /// Execute one claimed scheduled automation and return the terminal outcome.
    func execute(_ pendingRun: OperatorPendingScheduledRun) async throws -> OperatorAutomationExecutionOutcome
}

/// Cloud-backed automation executor that dispatches agent workflows via the control-plane.
public final class CloudOperatorAutomationExecutor: OperatorAutomationExecuting, @unchecked Sendable {
    private let controlPlaneBaseURL: () -> String
    private let credentialStore: CredentialStore
    private let preferences: () -> OperatorPreferences
    private let verificationReporter: OperatorVerificationReporting

    public init(
        controlPlaneBaseURL: @escaping () -> String,
        credentialStore: CredentialStore,
        preferences: @escaping () -> OperatorPreferences,
        verificationReporter: OperatorVerificationReporting
    ) {
        self.controlPlaneBaseURL = controlPlaneBaseURL
        self.credentialStore = credentialStore
        self.preferences = preferences
        self.verificationReporter = verificationReporter
    }

    public func execute(_ pendingRun: OperatorPendingScheduledRun) async throws -> OperatorAutomationExecutionOutcome {
        let correlationId = pendingRun.correlationId.isEmpty ? UUID().uuidString : pendingRun.correlationId
        let startedAt = ISO8601DateFormatter().string(from: Date())
        let baseUrl = controlPlaneBaseURL().trimmingCharacters(in: .whitespacesAndNewlines)

        guard !baseUrl.isEmpty else {
            let errorSummary = "Control-plane base URL is not configured."
            persistReport(
                correlationId: correlationId,
                pendingRun: pendingRun,
                startedAt: startedAt,
                terminalState: .errorNonRetryable,
                summary: errorSummary,
                providerId: nil,
                modelId: nil
            )
            return OperatorAutomationExecutionOutcome(
                state: .errorNonRetryable,
                summary: errorSummary,
                correlationId: correlationId,
                requiresAttention: true
            )
        }

        let prefs = preferences()
        let assignment = prefs.runtimeAssignments.assignment(for: .automation)
        let providerId = assignment.provider.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
            ? nil : assignment.provider
        let modelId = assignment.model.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
            ? nil : assignment.model
        let profileId = providerId.flatMap { prefs.activeProfileId(for: $0) } ?? "default"
        let apiKey = providerId.flatMap { credentialStore.providerApiKey(providerId: $0, profileId: profileId) }
        let providerBaseUrl = providerId.flatMap { prefs.activeProfile(for: $0)?.baseUrl }

        let request = AiWorkflowRequest(
            mode: "agent",
            provider: providerId ?? "",
            model: modelId,
            message: pendingRun.prompt,
            apiKey: apiKey,
            baseUrl: providerBaseUrl,
            correlationId: correlationId
        )

        let envelope: AiWorkflowJobEnvelope
        do {
            envelope = try await runAgentWorkflow(baseUrl: baseUrl, request: request)
        } catch {
            let errorSummary = error.localizedDescription
            persistReport(
                correlationId: correlationId,
                pendingRun: pendingRun,
                startedAt: startedAt,
                terminalState: .errorRetryable,
                summary: errorSummary,
                providerId: providerId,
                modelId: modelId
            )
            return OperatorAutomationExecutionOutcome(
                state: .errorRetryable,
                summary: errorSummary,
                correlationId: correlationId,
                requiresAttention: true
            )
        }

        let summary = resolveWorkflowSummary(envelope)
        let state = resolveWorkflowState(envelope)
        persistReport(
            correlationId: correlationId,
            pendingRun: pendingRun,
            startedAt: startedAt,
            terminalState: OperatorVerificationTerminalState(flowState: state),
            summary: summary,
            providerId: providerId,
            modelId: modelId
        )
        return OperatorAutomationExecutionOutcome(
            state: state,
            summary: summary,
            correlationId: correlationId,
            requiresAttention: state != .success
        )
    }

    // MARK: - Private

    private func runAgentWorkflow(
        baseUrl: String,
        request: AiWorkflowRequest
    ) async throws -> AiWorkflowJobEnvelope {
        guard let url = URL(string: "\(baseUrl)/api/ai/workflows/jobs") else {
            throw ControlPlaneClientError.invalidBaseURL(baseUrl)
        }
        var urlRequest = URLRequest(url: url)
        urlRequest.httpMethod = "POST"
        urlRequest.setValue("application/json", forHTTPHeaderField: "Content-Type")
        urlRequest.httpBody = try JSONEncoder().encode(request)

        let (data, _) = try await URLSession.shared.data(for: urlRequest)
        var envelope = try JSONDecoder().decode(AiWorkflowJobEnvelope.self, from: data)

        var pollCount = 0
        let maxPolls = ControlPlaneRuntimeConfig.shared.pollAttempts
        let pollInterval = ControlPlaneRuntimeConfig.shared.pollIntervalMs

        while !isTerminal(envelope) && pollCount < maxPolls {
            try await Task.sleep(nanoseconds: UInt64(pollInterval) * 1_000_000)
            guard let jobId = envelope.jobId,
                  let pollUrl = URL(string: "\(baseUrl)/api/ai/workflows/jobs/\(jobId)") else {
                break
            }
            let (pollData, _) = try await URLSession.shared.data(from: pollUrl)
            envelope = try JSONDecoder().decode(AiWorkflowJobEnvelope.self, from: pollData)
            pollCount += 1
        }
        return envelope
    }

    private func isTerminal(_ envelope: AiWorkflowJobEnvelope) -> Bool {
        guard let status = envelope.data?.status else { return true }
        return status == .succeeded || status == .failed || status == .cancelled
    }

    private func resolveWorkflowState(_ envelope: AiWorkflowJobEnvelope) -> FlowExecutionState {
        switch envelope.state {
        case .success: return .success
        case .errorRetryable: return .errorRetryable
        case .errorNonRetryable: return .errorNonRetryable
        case .unauthorized: return .unauthorized
        default:
            return envelope.data?.status == .succeeded ? .success : .errorNonRetryable
        }
    }

    private func resolveWorkflowSummary(_ envelope: AiWorkflowJobEnvelope) -> String {
        let reply = envelope.data?.result?.reply.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
        if !reply.isEmpty { return reply }
        let reason = envelope.data?.reason?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
        if !reason.isEmpty { return reason }
        return "Automation execution unavailable."
    }

    private func persistReport(
        correlationId: String,
        pendingRun: OperatorPendingScheduledRun,
        startedAt: String,
        terminalState: OperatorVerificationTerminalState,
        summary: String,
        providerId: String?,
        modelId: String?
    ) {
        let report = OperatorVerificationReport(
            correlationId: correlationId,
            scenario: .automationSchedule,
            runtime: OperatorVerificationRuntimeBinding(
                usage: .automation,
                transport: .cloud,
                provider: providerId,
                model: modelId
            ),
            triggerSource: .backgroundSchedule,
            request: OperatorVerificationRequestSummary(prompt: pendingRun.prompt),
            terminalState: terminalState,
            reply: OperatorVerificationReplySummary(
                message: summary,
                state: terminalState,
                provenance: "cloud"
            ),
            targetIds: OperatorVerificationTargetIds(scheduleId: pendingRun.scheduleId),
            startedAt: startedAt,
            completedAt: ISO8601DateFormatter().string(from: Date())
        )
        do {
            try verificationReporter.persist(report)
        } catch {
            // Verification report persistence is best-effort.
        }
    }
}
